### PR TITLE
AEA-3754 Use npm published versions of middy error handler and spine client in pfp AWS

### PR DESCRIPTION
**Pull request by kris-szlapa**
## Summary
🎫 [AEA-3754](https://nhsd-jira.digital.nhs.uk/browse/AEA-3754) **Use NPM Published Versions of Middy Error Handler and Spine Client in PfP AWS**

Once [AEA-3605](https://nhsd-jira.digital.nhs.uk/browse/AEA-3605) and [AEA-3617](https://nhsd-jira.digital.nhs.uk/browse/AEA-3617) are complete, we should change the AWS PfP code to use published npm packages for middy-error-handler and Github Packages for spineClient

- :exclamation: Breaking Change
- :robot: Operational or Infrastructure Change
- :sparkles: New Feature
- :warning: Potential issues that might be caused by this change

### Details

Use npm published versions of middy error handler and spine client in PfP AWS

The code for middy-error-handler and spineClient should be removed from the codebase and npm packages should be used instead

### CODE REVIEW

File: packages/getMyPrescriptions/tests/live-spine-client.test.ts [hunk 1]
```diff
@@ -112,11 +129,10 @@ describe("live spine client", () => {
     "throw error when $scenarioDescription",
     async ({httpResponseCode, spineStatusCode, nhsdLoginUser, errorMessage}) => {
       mock.onGet("https://spine/mm/patientfacingprescriptions").reply(httpResponseCode, {statusCode: spineStatusCode})
-      const spineClient = new LiveSpineClient(logger)
       const headers: APIGatewayProxyEventHeaders = {
         "nhsd-nhslogin-user": nhsdLoginUser
       }
-      await expect(spineClient.getPrescriptions(headers)).rejects.toThrow(errorMessage)
+      expect(() => extractNHSNumber(headers["nhsd-nhslogin-user"])).toThrow(errorMessage)
```
Comment (originalphil) [inline packages/getMyPrescriptions/tests/live-spine-client.test.ts:135]: We don't need the mocking in these test cases any more. We also don't need the httpResponseCode and spineStatusCode.
I'd suggest a simpler version of the type spineFailureTestData, with only the fields that we need.

File: .npmrc [hunk 2]
```diff
@@ -0,0 +1,2 @@
+//npm.pkg.github.com/:_authToken=${NODE_AUTH_TOKEN}
```
Comment (originalphil) [inline .npmrc:1]: Where do we get the token from, to install the package locally?

File: packages/capabilityStatement/src/capabilityStatement.ts [hunk 3]
```diff
@@ -47,4 +47,4 @@ export const handler = middy(lambdaHandler)
       }
     })
   )
-  .use(errorHandler({logger}))
+  .use(errorHandler({logger: console}))
```
Comment (MatthewPopat-NHS) [inline packages/capabilityStatement/src/capabilityStatement.ts:50]: Why did we change this?

File: package.json [hunk 4]
```diff
@@ -44,6 +42,8 @@
     "typescript": "^5.3.3"
   },
   "dependencies": {
+    "@nhs/fhir-middy-error-handler": "^1.0.2",
+    "@nhsdigital/eps-spine-client": "^1.0.1",
```
Comment (MatthewPopat-NHS) [inline package.json:46]: no need to add these I don't think

File: packages/getMyPrescriptions/src/getMyPrescriptions.ts [hunk 5]
```diff
@@ -67,6 +68,8 @@ const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayPro
         }
       }
     }
+    const nhsNumber = extractNHSNumber(event.headers["nhsd-nhslogin-user"])
+    logger.info(`nhsNumber: ${nhsNumber}`)
     const returnData = await spineClient.getPrescriptions(event.headers)
```
Comment (MatthewPopat-NHS) [inline packages/getMyPrescriptions/src/getMyPrescriptions.ts:73]: We need to add the nhsNumber to this set of headers

File: packages/getMyPrescriptions/tests/live-spine-client.test.ts [hunk 6]
```diff
@@ -72,7 +76,20 @@ describe("live spine client", () => {
...
+    "throw error when $scenarioDescription",
+    async ({httpResponseCode, spineStatusCode, nhsdLoginUser, errorMessage}) => {
+      mock.onGet("https://spine/mm/patientfacingprescriptions").reply(httpResponseCode, {statusCode: spineStatusCode})
+      const spineClient = new LiveSpineClient(logger)
+      const headers: APIGatewayProxyEventHeaders = {
+        "nhsd-nhslogin-user": nhsdLoginUser
+      }
+      await expect(spineClient.getPrescriptions(headers)).rejects.toThrow(errorMessage)
+    }
+  )
```
Comment (MatthewPopat-NHS) [inline packages/getMyPrescriptions/tests/live-spine-client.test.ts:90]: We don't need any of these spine client tests here, they now exist in the spine client repo. Just move the nhs number dependant ones below to extractNHSNumber.test.ts

File: packages/sandbox/src/sandbox.ts [hunk 7]
```diff
@@ -46,4 +46,4 @@ export const handler = middy(lambdaHandler)
       }
     })
   )
-  .use(errorHandler({logger}))
+  .use(errorHandler({logger: console}))
```
Comment (MatthewPopat-NHS) [inline packages/sandbox/src/sandbox.ts:49]: Why did we change this?

File: packages/statusLambda/src/statusLambda.ts [hunk 8]
```diff
@@ -56,4 +56,4 @@ export const handler = middy(lambdaHandler)
       }
     })
   )
-  .use(errorHandler({logger}))
+  .use(errorHandler({logger: console}))
```
Comment (MatthewPopat-NHS) [inline packages/statusLambda/src/statusLambda.ts:59]: Why did we change this?

File: packages/getMyPrescriptions/tests/live-spine-client.test.ts (hunk 1 above)
Comment (originalphil) [reply packages/getMyPrescriptions/tests/live-spine-client.test.ts:135]: Covered by removal of this test file, per Matt's comments.

File: packages/distanceSelling/tsconfig.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
     "rootDir": ".",
     "outDir": "lib"
   },
-  "references": [{"path": "../spineClient"}, {"path": "../common/testing"}],
+  "references": [{"path": "../common/testing"}],
```
[the remark this answers is not in the document]
Comment (originalphil) [reply packages/distanceSelling/tsconfig.json:7]: Please add serviceSearchClient as discussed.

File: package.json (hunk 4 above)
Comment (kris-szlapa) [reply package.json:46]: The dependencies have been removed

File: packages/getMyPrescriptions/src/getMyPrescriptions.ts (hunk 5 above)
Comment (kris-szlapa) [reply packages/getMyPrescriptions/src/getMyPrescriptions.ts:73]: nhsNumber has been added to the headers

File: packages/getMyPrescriptions/tests/live-spine-client.test.ts (hunk 6 above)
Comment (kris-szlapa) [reply packages/getMyPrescriptions/tests/live-spine-client.test.ts]: Duplicated tests have been removed

File: packages/sandbox/src/sandbox.ts (hunk 7 above)
Comment (kris-szlapa) [reply packages/sandbox/src/sandbox.ts]: The value `logger` has been assigned to the property `logger`
`.use(errorHandler({logger: logger}))`

File: packages/statusLambda/src/statusLambda.ts (hunk 8 above)
Comment (kris-szlapa) [reply packages/statusLambda/src/statusLambda.ts]: The value `logger` has been assigned to the property `logger`
`.use(errorHandler({logger: logger}))`

File: packages/capabilityStatement/src/capabilityStatement.ts (hunk 3 above)
Comment (kris-szlapa) [reply packages/capabilityStatement/src/capabilityStatement.ts]: The value `logger` has been assigned to the property `logger`
`.use(errorHandler({logger: logger}))`

File: .github/workflows/quality_checks.yml [hunk 10]
```diff
@@ -38,6 +38,14 @@ jobs:
         env:
           PYTHON_CONFIGURE_OPTS: --enable-shared 
 
+      - name: Install eps-spine-client package
```
[the remark this answers is not in the document]
Comment (kris-szlapa) [reply .github/workflows/quality_checks.yml:41]: Yes, if package-lock.json was generated with a valid token.

File: .github/workflows/sam_package_code.yml [hunk 11]
```diff
@@ -35,8 +36,16 @@ jobs:
         with:
           asdf_branch: v0.11.3
         env:
-          PYTHON_CONFIGURE_OPTS: --enable-shared 
-  
+          PYTHON_CONFIGURE_OPTS: --enable-shared
+
+      - name: Install eps-spine-client package
```
[the remark this answers is not in the document]
Comment (kris-szlapa) [reply .github/workflows/sam_package_code.yml:41]: `npm ci` works locally if `package-lock.json` was generated with a valid token.

File: .npmrc
Comment (MatthewPopat-NHS) [inline .npmrc:1]: if we're pulling in this file from the base machine for devs and populating it with a script for pipelines then there's no need to have it committed

File: .github/workflows/quality_checks.yml [hunk 12]
```diff
@@ -38,6 +38,14 @@ jobs:
...
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          echo "//npm.pkg.github.com/:_authToken=${NODE_AUTH_TOKEN}" >> ~/.npmrc
+          echo "@nhsdigital:registry=https://npm.pkg.github.com" >> ~/.npmrc
+          npm install @nhsdigital/eps-spine-client
```
Comment (MatthewPopat-NHS) [inline .github/workflows/quality_checks.yml:47]: shouldn't need to do this install, the make install below should do it for us

File: .npmrc
Comment (kris-szlapa) [reply .npmrc:1]: .npmrc file has been removed

File: packages/getMyPrescriptions/tests/live-spine-client.test.ts
Comment (originalphil) [inline packages/getMyPrescriptions/tests/live-spine-client.test.ts:1]: This file isn't needed any more, as the remaining tests are all in the spine client repo.

File: .github/workflows/sam_package_code.yml [hunk 13]
```diff
@@ -35,8 +36,16 @@ jobs:
...
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          echo "//npm.pkg.github.com/:_authToken=${NODE_AUTH_TOKEN}" >> ~/.npmrc
+          echo "@nhsdigital:registry=https://npm.pkg.github.com" >> ~/.npmrc
+          npm install @nhsdigital/eps-spine-client
```
[the remark this answers is not in the document]
Comment (kris-szlapa) [reply .github/workflows/sam_package_code.yml]: The install has been removed from `Setting up .npmrc` step

File: .github/workflows/quality_checks.yml (hunk 12 above)
Comment (kris-szlapa) [reply .github/workflows/quality_checks.yml]: The install has been removed from `Setting up .npmrc` step

File: packages/getMyPrescriptions/tests/live-spine-client.test.ts
Comment (kris-szlapa) [reply packages/getMyPrescriptions/tests/live-spine-client.test.ts:1]: Unnecessary tests have been removed.

**Review by originalphil — APPROVED** [review]

**Pull request merged**